Switch PyProject to Poetry

ppfeister opened the pull request:
Continuation of #2036

## Package info

Sherlock Project added as Author
Homepage URL set to the github[.]io homepage (PyPi homepage already set, but not pip show)
Dynamic versioning fixed (pulled from sherlock[.]py)
Maintainers added
Classifiers added
Keywords added
Description added
etc


___

__Refocused PR :__ PrProject being merged elsewhere to expedite package review process and resolve dependency issues (it takes time). This PR is now only to discuss switching to Poetry. (main pr #2116)

### ::: [>>> Jump to start of refocused discussion <<<](https://github.com/sherlock-project/sherlock/pull/2111#issuecomment-2113992558) :::

matheusfelipeog reviewed (commented):
Just a comment to start a discussion:

I personally use the Poetry package manager instead of directly using pip for projects like this. It has a slightly different format from what setuptools accepts; you can view it here: https://python-poetry.org/docs/pyproject/

I find its dependency resolution mechanism better, with features that prevent the addition of dependencies that are not compatible with others in the project programmatically.

Since in the future we intend to add Sherlock to PyPI, Poetry can also help with that, as it has native build and publish mechanisms, eliminating the need to use other tools like twine.

For the average user, nothing would change; pip would work normally.
For development, well, it has all these features I mentioned above and a few more.

What do you think about this?

---

Refs:

- #1241
- https://python-poetry.org/
- https://python-poetry.org/docs/basic-usage/
- https://python-poetry.org/docs/pyproject/
- https://python-poetry.org/docs/cli/

ppfeister commented:
Personally indifferent. No reservations or preference either way.
As long as normal pip usage isn't broken, then additional package manager support can only really be a plus.

As someone who hasn't used poetry, would this be as simple as making the pyproject compatible?

ppfeister commented:
So, I may be wrong on this, but it seems that poetry does not support dynamic versioning OOTB.

The package version will either have to be manually updated each time, or it would require an additional plugin that versions based on git tags (which sherlock doesn't currently use). Thoughts?

ppfeister commented:
It also doesn't seem that poetry has support for dynamic loading of dependencies -- they would need to be explicitly listed in the pyproj. Thoughts there?

Seems finicky on my end. I'm able to install with poetry right now, but unable to uninstall it with poetry (need to use pip). I'm also unable to run sherlock via `sherlock` due to an odd import error, needing to run it with `poetry run sherlock` instead. I'm sure both problems can be solved, just haven't seen why yet.

ppfeister commented:
Rebased onto master to fix conflict (some ~183 behind).

Poetry rewritten in a way that actually works.

Line 23 `packages = [ ...`  **will** likely present as a syntax error on some editors (i.e. those using the Even Better TOML plugin). This is NOT an error, and it is a feature added in version 1.8 of Poetry. There is another PyPI package named `sherlock`, and redirecting the install to `sherlock-project/sherlock` is another step at avoiding conflict.

Discovered that the current version of pyproject has some ~undesirable~ install issues, namely it doesn't install the src into a named directory and dumps them straight into site-packages for some unknown reason. Not only is this not pretty, it can cause some conflicts. This version fixes that behavior.

Due to how poetry handles packaging, the imports had to be rewritten.

Due to how the imports were rewritten, the (already poorly located) unit tests no longer function as desired.

Unit tests were relocated to the root where they probably should be anyways. Unit tests work from this location with modified imports.

PR action currently uses the *old* unit test location, so failure here **is expected**.

Unit tests will be fixed in another PR. I'll probably take advantage of that fixup to switch from the unittest module to tox, which is a bit more portable in my opinion.

This also removes the `requirements.txt` file, as dependencies should now be listed within the pyproj. If some sort of workflow requires the requirements.txt file, poetry itself can dump it from the pyproj.

Removed the setup.py and setup.cfg items as they are no longer necessary with the new layout.

Removed the rpm spec file.

Dynamic versioning still needs to be set up.

ppfeister commented:
(a couple minor doc updates included, i.e. pyproj readme logo and capitalization)

ppfeister commented:
Dynamic versioning added but not yet enabled. Relies on named tags and applies throughout Sherlock, rather than pulling from src.

ppfeister commented:
Fixed an import issue in entry point \_\_main\_\_.

Note that `python3 sherlock/sherlock` is not likely to work anymore, as that would require a bunch of try catch imports. `python3 -m sherlock` should be used to load the module properly now. Documentation updated to reflect.

```console
$ python -m sherlock --version
Sherlock  0.14.4
Requests  2.28.2
Python    3.12.3
```

matheusfelipeog reviewed (commented):
> So, I may be wrong on this, but it seems that poetry does not support dynamic versioning OOTB.
>
> The package version will either have to be manually updated each time, or it would require an additional plugin that versions based on git tags (which sherlock doesn't currently use). Thoughts?

Yes, poetry doesn't natively support dynamic versioning. We use community plugins for that. We don't need to rely on git tags; we can use the `__version__`  variable which should be located in `sherlock/__init__.py` (this is the location used by the community to store package metadata like version, name, etc.).

The plugin we can use for this is: https://github.com/tiangolo/poetry-version-plugin

> It also doesn't seem that poetry has support for dynamic loading of dependencies -- they would need to be explicitly listed in the pyproj. Thoughts there?

This isn't necessary. They should be defined directly in the dependencies group of the project in `pyproject.toml`. I see that you've already done that, good job.

> Seems finicky on my end. I'm able to install with poetry right now, but unable to uninstall it with poetry (need to use pip). I'm also unable to run sherlock via sherlock due to an odd import error, needing to run it with poetry run sherlock instead. I'm sure both problems can be solved, just haven't seen why yet.

Poetry is a package manager like `pip` and also an environment manager like `virtualenv`. To use it correctly, you need to activate the virtual environment by using `poetry shell` in the project directory.

The command `poetry run script.py` does exactly that but does not keep the virtual environment active in the shell session. The virtual environment will only be used to run the script and then it will exit.

ppfeister commented:
Most of my original comments seem to be resolved. It was easier to just _rewrite_ it rather than adapt it for some reason.

> Yes, poetry doesn't natively support dynamic versioning. We use community plugins for that. We don't need to rely on git tags; we can use the __version__ variable which should be located in sherlock/__init__.py (this is the location used by the community to store package metadata like version, name, etc.).
> 
> The plugin we can use for this is: https://github.com/tiangolo/poetry-version-plugin

My original find was `poetry-dynamic-versioning`, but that dynamically populates version tags within the package rather than dynamically populating the pyproj from tags within the package (backwards). Your linked plugin might be better since it pulls from src as before

Only problem I've noticed is that there doesn't seem to be a way to *require* your plugin for builds. Since there is no backend, it also doesn't change how things are built sans-`poetry build`. python -m build will now forever use version 0, and poetry build itself will return 0 by default unless a dev goes out of their way to realize it's needed and install the plugin.

Any suggestions to enforce this plugin, or just let people figure it out?

matheusfelipeog commented:
There is no way to require poetry plugins because they are not part of the project itself but rather part of poetry. This is a community need, and there are discussions about implementing this in the future.

Ideally, we should have a [workflow for build and publish](https://github.com/search?q=poetry-version-plugin+path%3A%2F%5E.github%5C%2Fworkflows%5C%2F%2F&type=code) so we don't handle this manually.

For those who want to build locally from the source for any reason, this should be documented in the respective section of the documentation.

ppfeister commented:
Well in that case, that's all I've got. _Probably_ good to merge...

matheusfelipeog reviewed (changes requested):
Here are a few more suggestions.

ppfeister commented on pyproject.toml:
```diff
@@ -1,48 +1,56 @@
-    { name = "Matheus Felipe" },
-    { name = "Sondre Karlsen Dyrnes" },
-    { name = "Paul Pfeister" }
+    "Paul Pfeister",
+    "Matheus Felipe",
+    "Sondre Karlsen Dyrnes"
```
Believe it or not, this breaks PyPI.
It seems that you can have **either** multiple maintainers **or** a maintainer with an email address. When you have multiple with email addresses, it only shows the first maintainer for some unknown reason.

ppfeister replied:
Seems that Poetry does that _anyways_ is PyPI so the point I raised is moot

ppfeister commented on pyproject.toml:
```diff
@@ -1,48 +1,56 @@
+license = "MIT"
 authors = [
-    { name = "Siddharth Dushantha" }
+    "Siddharth Dushantha"
```
This was a choice made by `@sdushantha`. I support having email addresses but he was wary of spam.

What do you think @sdushantha

sdushantha replied:
Just noticed that my email address is in every commit. So we may as well just include it in the `pyproject.toml`

ppfeister commented on pyproject.toml:
```diff
@@ -1,48 +1,56 @@
     "Programming Language :: Python :: 3",
     "Topic :: Security"
 ]
+homepage = "http://sherlock-project.github.io/"
```
:man_facepalming:

ppfeister commented on pyproject.toml:
```diff
@@ -1,48 +1,56 @@
     "Topic :: Security"
 ]
+homepage = "http://sherlock-project.github.io/"
+repository = "https://github.com/sherlock-project/sherlock.git"
```
I've always been torn on this. I like the `.git` when referring to the repo itself (rather than the web page for the repo) for some extra verbosity, but it's very not necessary

I support removal

ppfeister commented on pyproject.toml:
```diff
@@ -1,48 +1,56 @@
-[tool.setuptools]
-package-dir = {"" = "sherlock"}
+[tool.poetry.dependencies]
+python = "^3.8"
+certifi = ">=2019.6.16"
+colorama = ">=0.4.1"
+PySocks = ">=1.7.0"
+requests = ">=2.22.0"
+requests-futures = ">=1.0.0"
+stem = ">=1.8.0"
+torrequest = ">=0.1.0"
+pandas = ">=1.0.0"
+openpyxl = ">=3.0.10"
+exrex = ">=0.11.0"
```
I believe the original requirements.txt used raw greater-thans as well. When I copied them over and changed them to carets, things broke as the min version was so out of date and no longer available.

Let's see if I can grab current versions that are available on all targets... Will update here.

ppfeister replied:
Done

sdushantha commented on sherlock/sherlock.py:
```diff
@@ -19,18 +19,26 @@
 
 import requests
 
+# Removing __version__ here will trigger update message for users
+# Do not remove until ready to trigger that message
+__version__ = "0.14.4"
+del __version__
```
Since this variable is only being used for checking for updates, cant we just remove this and then check the `__init__,py` file instead?

ppfeister replied:
We could (and I forgot to make that change, so good call), but we still have to leave that text in place until ready to bump the version

If we remove it from sherlock.py now, before the next version num is ready, it'll trigger unnecessary update available notifications

Will commit a change to pull version num from __init__ in a bit

ppfeister replied:
@sdushantha Changed url to check __init__
Note that testing this feature locally will fail since it doesn't actually exist til merged

sdushantha reviewed (approved):
LGMT
